sched-simple: fix alloc/free for resource sets with equal size but different cpusets

grondo opened the pull request:
This PR fixes #2202. The problem was a dumb implementation of an idset comparator internal to `sched-simple`'s `rlist` object. The `idset_cmp()` function compared non-equal idsets using `idset_count()`, but of course two idsets can have the same count but still not be equivalent.

The new implementation detects unequal idsets that have the same "count" and then falls back to comparing the first different stored integer.

This appears to fix the weirdness under Slurm with cpu-affinity:

```
$ srun --pty -N32 -n128 src/cmd/flux start
$ 
$ src/modules/sched-simple/rlist-query 
rank[2,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,66,70,74,78,82,86,90,94,98,102,106,110,114,118,122,126]/core2 rank[1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61,65,69,73,77,81,85,89,93,97,101,105,109,113,117,121,125]/core1 rank[3,7,11,15,19,23,27,31,35,39,43,47,51,55,59,63,67,71,75,79,83,87,91,95,99,103,107,111,115,119,123,127]/core3 rank[0,4,8,12,16,20,24,28,32,36,40,44,48,52,56,60,64,68,72,76,80,84,88,92,96,100,104,108,112,116,120,124]/core0
$ flux srun -N128 hostname
$ echo $?
0
$ flux dmesg | tail
2019-06-21T16:47:13.705278Z sched-simple.debug[0]: req: 1069547520000: spec={128,128,1}
2019-06-21T16:47:13.705713Z sched-simple.debug[0]: alloc: 1069547520000: rank[0,4,8,12,16,20,24,28,32,36,40,44,48,52,56,60,64,68,72,76,80,84,88,92,96,100,104,108,112,116,120,124]/core0 rank[1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61,65,69,73,77,81,85,89,93,97,101,105,109,113,117,121,125]/core1 rank[3,7,11,15,19,23,27,31,35,39,43,47,51,55,59,63,67,71,75,79,83,87,91,95,99,103,107,111,115,119,123,127]/core3 rank[2,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,66,70,74,78,82,86,90,94,98,102,106,110,114,118,122,126]/core2
2019-06-21T16:47:13.718219Z kvs.debug[0]: aggregated 2 transactions (2 ops)
2019-06-21T16:47:13.723366Z sched-simple.debug[0]: free: rank[0,4,8,12,16,20,24,28,32,36,40,44,48,52,56,60,64,68,72,76,80,84,88,92,96,100,104,108,112,116,120,124]/core0 rank[1,5,9,13,17,21,25,29,33,37,41,45,49,53,57,61,65,69,73,77,81,85,89,93,97,101,105,109,113,117,121,125]/core1 rank[3,7,11,15,19,23,27,31,35,39,43,47,51,55,59,63,67,71,75,79,83,87,91,95,99,103,107,111,115,119,123,127]/core3 rank[2,6,10,14,18,22,26,30,34,38,42,46,50,54,58,62,66,70,74,78,82,86,90,94,98,102,106,110,114,118,122,126]/core2

chu11 commented on src/modules/sched-simple/rlist.c:
```diff
@@ -85,9 +85,24 @@ static struct rnode *rlist_find_rank (struct rlist *rl, uint32_t rank)
+        unsigned int b = idset_first (set2);
+        while ((rv = (a - b)) == 0) {
+            a = idset_next (set1, a);
+            b = idset_next (set1, b);
```
should be `idset_next (set2)`?

grondo replied:
yep, thanks!

grondo replied:
might have to quickly figure out how to ensure that code gets exercised, but I may not get to it...

grondo replied:
@chu11, quickly fixed this issue with a new commit. Once you're done reviewing, I'll squash.

chu11 commented on src/modules/sched-simple/test/rlist.c:
```diff
@@ -202,6 +202,74 @@ static void test_simple (void)
+    char *result = NULL;
+    struct rlist *a = NULL;
+
+	struct rlist *rl = rlist_from_hwloc_by_rank (R_issue2202);
```
nit: used tabs instead of spaces

grondo replied:
ugh, sorry about that. Somehow the editor I was using seems to have ignored modeline...

chu11 commented on src/modules/sched-simple/test/rlist.c:
```diff
@@ -202,6 +202,74 @@ static void test_simple (void)
+        rlist_destroy (a);
+    }
+        
+	rlist_destroy (rl);
```
nit: used tabs instead of spaces

chu11 commented on src/modules/sched-simple/test/rlist.c:
```diff
@@ -249,6 +317,7 @@ int main (int ac, char *av[])
     run_test_entries (test_2n_4c,       2, 4);
     run_test_entries (test_6n_4c,       6, 4);
     run_test_entries (test_1024n_4c, 1024, 4);
+	test_issue2202 ();
```
nit: used tabs instead of spaces

grondo commented:
I'll also work on getting better coverage by adding another testcase for ranks with cpusets with more than one core. That should exercise the idset_cmp() loop and get better coverage (and would have likely found the typo @chu11 discovered earlier)

chu11 commented:
things look good to me, squash away

grondo commented:
Thanks, squashed and added some extra tests to ensure the inner loop of `idset_cmp` is exercised. (I was going pretty quick here, I'll check back after lunch and give it another once over once I see coverage results)

grondo commented:
@chu11, actually I had a change of heart on `idset_cmp()`. It didn't seem logical to first sort sets based on `idset_count()` *then* the first non-matching integer id. This would sort the sets `"0-1","10"` as `"10", "0-1"`. So instead I just changed `idset_cmp()` to first compare the first non-matching id from each set, where the smaller of the two sets would sort "first" if the N ids of each set match.

Sorry, but can you review again? (You had some great finds last time)

I went ahead and squashed just in case the current changes are good and can be merged right away.

chu11 commented:
@grondo looks good to me.  Are you all done?

grondo commented:
> @grondo looks good to me. Are you all done?

Yep, thanks!!

grondo commented:
Thanks!
